File: deno-fetch/crawl.js
import { readFileSync } from 'node:fs';
import { PromisePool } from '@supercharge/promise-pool'

const CONCURRENCY = parseInt(Deno.env.get('CONCURRENCY') || 10)
const REQUEST_TIMEOUT = parseInt(Deno.env.get('REQUEST_TIMEOUT') || 5)
const filePath = Deno.args[0] || 'data/top-1000.txt'
const file = readFileSync(filePath, 'utf8')
const urls = file.split(/\n/)
const start = new Date()

console.log(`Starting crawl with ${CONCURRENCY} concurrency`)
console.log(`Request timeout: ${REQUEST_TIMEOUT}s`)
console.log(`Reading URLs from ${filePath}`)

const headers = {
    'User-Agent': 'crawler-test',
    'Accept-Encoding': 'gzip, deflate, br'
}

const makeRequest = async url => {
    const startTime = Date.now()

    TypeError

    try {
        const controller = new AbortController();
        const signal = controller.signal;
        const timeout = setTimeout(() => controller.abort('TIMEOUT_ERROR'), 5000);
        const response = await fetch(url, { headers, signal })
        clearTimeout(timeout);
        const time = Date.now() - startTime
        console.log(`${url}: ${response.status} -- ${time}ms`)
        return { code: response.status, time }
    } catch (error) {
        const time = Date.now() - startTime
        const code = error.name || error
        console.error(`${url}: ${code} -- ${time}ms`)
        return { code, time }
    }
}

const { results, errors } = await PromisePool
    .for(urls)
    .withConcurrency(CONCURRENCY)
    .process(makeRequest)

const aggregates = results.reduce((agg, result) => {
    return { ...agg, [result.code]: (agg[result.code] || 0) + 1 }
}, {})
const avgTime = results.reduce((agg, result) => {
    return agg + result.time
}, 0) / results.length
const medianTime = results.map(r => r.time).sort()[Math.floor(results.length / 2)]

console.log(`Total time: ${(Date.now() - start) / 1000}s`)
console.log(`Average time: ${avgTime}`)
console.log(`Median time: ${medianTime}`)
console.log(`Total URLs: ${Object.values(aggregates).reduce((agg, count) => agg + count, 0)}`)

console.log(aggregates)

File: rust-reqwest/crawl.rs
use std::time::{Duration, SystemTime};
use std::fs::File;
use std::env;
use std::io::{self, prelude::*, BufReader};
use reqwest;
use reqwest::header::USER_AGENT;
use tokio;
use futures::prelude::*;
use std::collections::HashMap;

struct Result {
    code: String,
    time: Duration
}


fn get_env(key: &str, default: u16) -> u16 {
    env::var(key)
        .ok()
        .and_then(|limit| limit.parse::<u16>().ok())
        .unwrap_or(default)
}

#[tokio::main]
async fn main() -> io::Result<()> {
    let url_limit = get_env("LIMIT", 1000);
    let request_timeout = get_env("REQUEST_TIMEOUT", 5);
    let concurrency = get_env("CONCURRENCY", 10);

    println!("Starting crawl:");
    println!(" * {}: {:?}", "URL_LIMIT", url_limit);
    println!(" * {}: {:?}", "REQUEST_TIMEOUT", request_timeout);
    println!(" * {}: {:?}", "CONCURRENCY", concurrency);

    let time = SystemTime::now();

    let file = File::open("data/urls.txt")?;
    let reader = BufReader::new(file);

    let results = stream::iter(reader.lines().take(url_limit as usize))
        .map(|line| {
            async move {
                let start = SystemTime::now();
                let url = line.unwrap();

                let client = reqwest::Client::builder()
                    .timeout(Duration::from_secs(request_timeout as u64))
                    .build()
                    .unwrap();

                let response = client.get(&url)
                    .header(USER_AGENT, "crawler-test")
                    .header("Accept-Encoding", "gzip, deflate, br")
                    .send()
                    .await;

                match response {
                    Ok(response) => {
                        let time = start.elapsed().unwrap();
                        println!("{}: {} -- {:?}", url, response.status(), time);
                        Result {
                            code: response.status().to_string(),
                            time
                        }
                    }
                    Err(err) => {
                        let time = start.elapsed().unwrap();
                        println!("{}: {} -- {:?}", url, err, time);
                        Result {
                            code: err.without_url().to_string(),
                            time
                        }
                    }
                }
            }
        })
        .buffer_unordered(concurrency as usize)
        .collect::<Vec<Result>>()
        .await;


    let aggregates = results.iter().fold(HashMap::new(), |mut acc, result| {
        *acc.entry(result.code.clone()).or_insert(0) += 1;
        acc
    });

    let mut sorted_aggregates = aggregates.iter().collect::<Vec<(&String, &u16)>>();

    sorted_aggregates.sort_by(|a, b| b.1.cmp(&a.1));

    for (code, count) in &mut sorted_aggregates {
        println!("{}: {}", code, count);
    }

    let total_time = time.elapsed().unwrap();
    let total_urls = aggregates.values().sum::<u16>();
    let mut sorted_times = results.iter().map(|result| result.time).collect::<Vec<Duration>>();
    sorted_times.sort_by(|a, b| b.cmp(a));
    let avg_time = sorted_times.iter().sum::<Duration>() / total_urls as u32;
    let median_time = sorted_times[sorted_times.len() / 2];

    println!("Total time: {:?}", total_time);
    println!("Total URLs: {:?}", total_urls);
    println!("Average time: {:?}", avg_time);
    println!("Median time: {:?}", median_time);

    Ok(())
}